src/core/eval.ml

```diff
@@ -212,7 +214,18 @@ and tree_walk : dtree -> ctxt -> stack -> (term * stack) option =
         in
         List.iter fn env_builder;
         (* Actually perform the action. *)
-        Some(Bindlib.msubst act env, stk)
+        begin
+          try Some(Bindlib.msubst act env, stk)
+          with Invalid_argument(_) ->
```

Comment:
It looks kind of hackish to rely on an Invalid_argument exception... Isn't there another way to do this? Anyway, a comment is necessary.

Reply:
Yes, we can as well check whether the `xvars` is equal to 0. Performance wise, I thought that relying on the exception would avoid a comparison in the case of regular rewriting rules. I'll check if there's an actual difference.